pkg/application/modify.go

```diff
@@ -29,51 +29,41 @@ func (app *Application) SetTrait(componentName, traitType string, traitData map[
 	if traitData == nil {
 		traitData = make(map[string]interface{})
 	}
-	if app.Components == nil {
-		app.Components = make(map[string]map[string]interface{})
+
+	s, ok := app.Services[componentName]
+	if !ok {
+		s = appfile.Service{}
 	}
-	comp := app.Components[componentName]
-	if comp == nil {
-		comp = make(map[string]interface{})
+
+	t, ok := s[traitType]
+	if !ok {
+		t = make(map[string]interface{})
 	}
-	traits, err := app.GetTraits(componentName)
-	if err != nil {
-		return err
+	tm := t.(map[string]interface{})
+	for k, v := range traitData {
+		tm[k] = v
 	}
-	traits[traitType] = traitData
-	comp[Traits] = traits
-	app.Components[componentName] = comp
-	return app.Validate()
+	return nil
 }
 
 func (app *Application) RemoveTrait(componentName, traitType string) error {
 	if app == nil {
 		return errors.New("app is nil pointer")
 	}
-	if app.Components == nil {
-		app.Components = make(map[string]map[string]interface{})
-	}
-	comp := app.Components[componentName]
-	if comp == nil {
-		comp = make(map[string]interface{})
-	}
-	traits, err := app.GetTraits(componentName)
-	if err != nil {
-		return err
+
+	s, ok := app.Services[componentName]
+	if !ok {
+		return nil
 	}
-	delete(traits, traitType)
-	comp[Traits] = traits
-	app.Components[componentName] = comp
-	return app.Validate()
+	delete(s, traitType)
```

Comment:
we should put it back into the map, or it won't work

Reply:
Why? I think the map is a pointer and the deletion happens in place

Reply:
See https://play.golang.org/p/PfoAtCmdCrh